[CoordinatedGraphics] Provide scheduled display-update events for better scrolling handling

zdobersek opened the pull request:
#### 06b9ec1834d27a4fb8a324e29908ed274305620a
<pre>
[CoordinatedGraphics] Provide scheduled display-update events for better scrolling handling
<a href="https://bugs.webkit.org/show_bug.cgi?id=252467">https://bugs.webkit.org/show_bug.cgi?id=252467</a>

Reviewed by Carlos Garcia Campos.

In ThreadedCompositor, a timer is established on the composition thread that is
used to mimic display updates at the desired update frequency. This is necessary
in order to provide opportunities for update producers like the scrolling thread
to actually enact the desired composition changes, like scrolling, independently
of the main thread.

The timer is put on the composition thread, matching the compositing-thread
update timer in priority. Its delay is based on the established frequency of the
display updates.

ThreadedCompositor::displayUpdateFired() method is added and is dispatched upon
every firing of the timer. It notifies the scrolling thread(s) for the given
display ID through the EventDispatcher instance on the WebProcess singleton,
and the next firing of the timer is then scheduled.

When ThreadedCompositor enters into the composition phase, the display update
timer is stopped, and the displayUpdateFired() dispatcher is invoked manually
when the frame-complete signal is returned from the embedder&apos;s environment. This
allows aligning the display update timer with the environment&apos;s vsync signals
without disturbing the composition itself.

The target update frequency logic is pulled from ThreadedDisplayRefreshMonitor
into ThreadedCompositor, with the desired DisplayUpdate object now also kept in
that class. In order for ThreadedDisplayRefreshMonitor to use a sensible
DisplayUpdate object for each of its own dispatches, the DisplayUpdate object
is passed through the requiresDisplayRefreshCallback() method to the monitor,
where it&apos;s then stored for next invocation on the main thread, if it occurs.

ScrollingTreeNicosia::applyLayerPositionsInternal() override is provided,
establishing an update scope when running on the scrolling thread. This will
cause the applied positions to actually go through composition whenever the
scrolling is being handled on the scrolling thread, meaning outside of the usual
update process.

* Source/WebCore/page/scrolling/nicosia/ScrollingTreeNicosia.cpp:
(WebCore::ScrollingTreeNicosia::applyLayerPositionsInternal):
* Source/WebCore/page/scrolling/nicosia/ScrollingTreeNicosia.h:
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp:
(WebKit::m_displayRefreshMonitor):
(WebKit::ThreadedCompositor::renderLayerTree):
(WebKit::ThreadedCompositor::sceneUpdateFinished):
(WebKit::ThreadedCompositor::frameComplete):
(WebKit::ThreadedCompositor::targetRefreshRateDidChange):
(WebKit::ThreadedCompositor::displayUpdateFired):
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.h:
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedDisplayRefreshMonitor.cpp:
(WebKit::ThreadedDisplayRefreshMonitor::ThreadedDisplayRefreshMonitor):
(WebKit::ThreadedDisplayRefreshMonitor::requiresDisplayRefreshCallback):
(WebKit::ThreadedDisplayRefreshMonitor::invalidate):
(WebKit::ThreadedDisplayRefreshMonitor::displayRefreshCallback):
(WebKit::ThreadedDisplayRefreshMonitor::setTargetRefreshRate): Deleted.
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedDisplayRefreshMonitor.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::displayDidRefresh):

Canonical link: <a href="https://commits.webkit.org/261124@main">https://commits.webkit.org/261124@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/38175827b4dfcb2520afdede0e9a48fa35c9aa87

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110679 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19760 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/43243 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/2029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119515 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114647 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21169 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10872 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/1540 "layout-tests (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116420 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15766 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/43243 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102976 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/43243 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44086 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/12396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/43243 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/85982 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12945 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18317 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/43243 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7706 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/14883 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp:
```diff
@@ -300,13 +311,26 @@ WebCore::DisplayRefreshMonitor& ThreadedCompositor::displayRefreshMonitor() cons
+{
+    m_display.displayUpdate = m_display.displayUpdate.nextUpdate();
+
+    WebProcess::singleton().eventDispatcher().notifyScrollingTreesDisplayDidRefresh(m_display.displayID);
```
Why did you remove the `Client::displayDidRefresh`? `EventDispatcher::notifyScrollingTreesDisplayDidRefresh` iterates all scrolling trees to notify that a display has been refreshed. Every tree checks if its displayID matches the given one. This is required when there's a global display notifying about a refresh, but in our case we have one display refresh monitor with its own displayID for every threaded compositor. So, we know what tree we wat to notify without having to iterate all trees.

zdobersek replied:
This is called from the composition thread, so `LayerTreeHost::displayDidRefresh()` ends up loading the Page's ScrollingCoordinator and then its ScrollingTree from a non-main thread.

Using EventDispatcher avoids that, and it locks over the scrolling trees that are registered and unregistered on the main thread, so no threading problems.

carlosgcampos replied:
I see, in that case I would add:

```
void EventDispatcher::notifyWebPageScrollingTreeDisplayDidRefresh(WebPage& page, PlatformDisplayID displayID);
```

And call it from `LayerTreeHost::displayDidRefresh`. This avoids the tree iteration, fixes the threading issues and doesn't use WebProcess API from Shared.

zdobersek replied:
I assume you would want to short-cut to the ScrollingTree in the new EventDispatcher method, through the WebPage. But that still operates on that WebPage on a non-main thread, so threading problems would remain.

Calling `ThreadedScrollingTree::displayDidRefresh()` on every scrolling tree isn't ideal, and it can be improved, but I don't see it as critical.

carlosgcampos replied:
It would be just getting the identifier that is the key of the trees map. Anyway, it's not critical since in most of the cases there's only one page per process. But still I prefer to call the dispatcher from `LayerTreeHost::displayDidRefresh`, which is in WebProcess layer too. At some point we should move the threaded compositor code from Shared to WebProcess and stop pretending it's shared, though.

zdobersek replied:
Yes, I'm adding back `LayerTreeHost::displayDidRefresh`.

carlosgcampos commented on Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp:
```diff
@@ -64,7 +66,17 @@ ThreadedCompositor::ThreadedCompositor(Client& client, ThreadedDisplayRefreshMon
+    m_display.displayUpdate = { 0, c_defaultRefreshRate / 1000 };
+
     m_compositingRunLoop->performTaskSync([this, protectedThis = Ref { *this }] {
+        m_display.updateTimer = makeUnique<RunLoop::Timer>(RunLoop::current(), this, &ThreadedCompositor::displayUpdateFired);
```
I think we should delete the timer in `invalidate()`, inside the performTaskSync to ensure it happens in the thread it was created.

zdobersek replied:
OK.

carlosgcampos commented on Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp:
```diff
@@ -300,13 +310,26 @@ WebCore::DisplayRefreshMonitor& ThreadedCompositor::displayRefreshMonitor() cons
+
+    if (!rate)
+        rate = c_defaultRefreshRate;
+    m_display.displayUpdate = { 0, rate / 1000 };
```
I see displayUpdate is always used in the compositing thread except here. Should we use `m_compositingRunLoop->performTask` to do this in the compositing thread instead?

zdobersek replied:
Makes sense, `performTaskSync`.

carlosgcampos reviewed (approved):
I still get some blocks, but less often, so let's try this. Thanks!

zdobersek committed:
[CoordinatedGraphics] Provide scheduled display-update events for better scrolling handling
https://bugs.webkit.org/show_bug.cgi?id=252467

Reviewed by Carlos Garcia Campos.

In ThreadedCompositor, a timer is established on the composition thread that is
used to mimic display updates at the desired update frequency. This is necessary
in order to provide opportunities for update producers like the scrolling thread
to actually enact the desired composition changes, like scrolling, independently
of the main thread.

The timer is put on the composition thread, matching the compositing-thread
update timer in priority. Its delay is based on the established frequency of the
display updates.

ThreadedCompositor::displayUpdateFired() method is added and is dispatched upon
every firing of the timer. It notifies the scrolling thread(s) for the given
display ID through the EventDispatcher instance on the WebProcess singleton,
and the next firing of the timer is then scheduled.

When ThreadedCompositor enters into the composition phase, the display update
timer is stopped, and the displayUpdateFired() dispatcher is invoked manually
when the frame-complete signal is returned from the embedder's environment. This
allows aligning the display update timer with the environment's vsync signals
without disturbing the composition itself.

The target update frequency logic is pulled from ThreadedDisplayRefreshMonitor
into ThreadedCompositor, with the desired DisplayUpdate object now also kept in
that class. In order for ThreadedDisplayRefreshMonitor to use a sensible
DisplayUpdate object for each of its own dispatches, the DisplayUpdate object
is passed through the requiresDisplayRefreshCallback() method to the monitor,
where it's then stored for next invocation on the main thread, if it occurs.

ScrollingTreeNicosia::applyLayerPositionsInternal() override is provided,
establishing an update scope when running on the scrolling thread. This will
cause the applied positions to actually go through composition whenever the
scrolling is being handled on the scrolling thread, meaning outside of the usual
update process.

* Source/WebCore/page/scrolling/nicosia/ScrollingTreeNicosia.cpp:
(WebCore::ScrollingTreeNicosia::applyLayerPositionsInternal):
* Source/WebCore/page/scrolling/nicosia/ScrollingTreeNicosia.h:
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp:
(WebKit::m_displayRefreshMonitor):
(WebKit::ThreadedCompositor::renderLayerTree):
(WebKit::ThreadedCompositor::sceneUpdateFinished):
(WebKit::ThreadedCompositor::frameComplete):
(WebKit::ThreadedCompositor::targetRefreshRateDidChange):
(WebKit::ThreadedCompositor::displayUpdateFired):
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.h:
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedDisplayRefreshMonitor.cpp:
(WebKit::ThreadedDisplayRefreshMonitor::ThreadedDisplayRefreshMonitor):
(WebKit::ThreadedDisplayRefreshMonitor::requiresDisplayRefreshCallback):
(WebKit::ThreadedDisplayRefreshMonitor::invalidate):
(WebKit::ThreadedDisplayRefreshMonitor::displayRefreshCallback):
(WebKit::ThreadedDisplayRefreshMonitor::setTargetRefreshRate): Deleted.
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedDisplayRefreshMonitor.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::displayDidRefresh):

Canonical link: https://commits.webkit.org/261124@main
